### PR TITLE
feat(storage): wake up memory acquire request in order

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/15786
See design details in issue

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/utils.rs [hunk 1]
```diff
@@ -166,26 +167,47 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
     }
 
+    fn may_notify_waiters(self: &Arc<Self>) {
+        let mut waiters = self.controller.lock();
+        while let Some((tx, quota)) = waiters.pop_front() {
+            if !self.try_require_memory_in_capacity(quota, self.quota) {
+                waiters.push_front((tx, quota));
+                break;
+            }
+            let _ = tx.send(MemoryTrackerImpl::new(self.clone(), quota));
```
Comment (wenym1) [inline src/storage/src/hummock/utils.rs:196]: IIUC, the difference between `MemoryTracker` and `MemoryTrackerImpl` is that, `MemoryTrackerImpl` only releases the quota, but `MemoryTracker` will also notify waiters. 

If so, I've come up with a corner case that may cause endless await.

Let's say the capacity is 100, and we have two threads, A wants 100 and B wants 100, and both of them are currently in the queue. 

1. When someone else releases the quota and call `may_notify_waiters`, it will find A in the queue, and create a `MemoryTrackerImpl` with quota 100 and send to A via the tx. The `send` method returns successfully and the `MemoryTrackerImpl` is added to the buffer. 
2. However, before A poll the rx and take the `MemoryTrackerImpl` out, the future of A may be cancelled and dropped, and then only the `MemoryTrackerImpl` gets dropped, which only adds the quota back to the atomic variable, but won't notify waiters in the queue (which is B). 
3. By now, B is waiting for notification, but no one is going to release any quota and notify B, which causes endless await.

File: src/storage/src/hummock/utils.rs [hunk 2]
```diff
@@ -310,41 +338,57 @@ impl MemoryLimiter {
...
+
+// We must notify waiters outside `MemoryTrackerImpl` to avoid dead-lock and loop-owner.
 impl Drop for MemoryTracker {
     fn drop(&mut self) {
-        self.limiter.release_quota(self.quota);
+        if let Some(quota) = self.inner.quota.take() {
+            self.inner.limiter.release_quota(quota);
+        }
+        // check `inflight_barrier` to avoid access lock every times drop `MemoryTracker`.
+        if self
```
Comment (wenym1) [inline src/storage/src/hummock/utils.rs:383]: Maybe we can optimize the case when there are concurrent drops in multiple threads. 

A rough idea will be, each thread will try to set the `pending_request_count` to 0 via CAS in a loop, and the only one that successfully sets it to 0 will do the notification.

File: src/common/src/config.rs [hunk 3]
```diff
@@ -721,6 +721,9 @@ pub struct StorageConfig {
     #[serde(default = "default::storage::mem_table_spill_threshold")]
     pub mem_table_spill_threshold: usize,
 
+    #[serde(default = "default::storage::memory_limit_for_behind_barrier_ratio")]
+    pub memory_limit_for_behind_barrier_ratio: u64,
```
Comment (wenym1) [inline src/common/src/config.rs:725]: The naming is confusing. From the code, it might mean `memory_limiter_fast_quota_ratio`?

**Review by wenym1 — COMMENTED** [review]
The current memory limiter is very similar to [tokio::sync::Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html). Maybe we can try borrowing some implementation details from it.

File: src/storage/src/hummock/utils.rs [hunk 4]
```diff
@@ -203,93 +225,99 @@ impl MemoryLimiterInner {
...
-                        }
-                    }
+    async fn require_memory(self: &Arc<Self>, quota: u64) -> MemoryTrackerImpl {
+        let waiter = {
+            let mut waiters = self.controller.lock();
+            if waiters.is_empty() {
+                // When this request is the first waiter but the previous `MemoryTracker` is just release a large quota, it may skip notifying this waiter because it has checked `inflight_barrier` and found it was zero. So we must set it one and retry `try_require_memory_in_capacity` again to avoid deadlock.
+                self.pending_request_count.store(1, AtomicOrdering::Release);
+                if self.try_require_memory_in_capacity(quota, self.fast_quota) {
+                    self.pending_request_count.store(0, AtomicOrdering::Release);
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:238]: This method is always called before `try_require_memory`, it means that has a failed try by CAS without lock

File: src/storage/src/hummock/utils.rs [hunk 5]
```diff
@@ -166,26 +167,47 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
-        self.notify.notify_waiters();
     }
 
     fn add_memory(&self, quota: u64) {
         self.total_size.fetch_add(quota, AtomicOrdering::SeqCst);
     }
 
+    fn may_notify_waiters(self: &Arc<Self>) {
+        let mut waiters = self.controller.lock();
+        while let Some((tx, quota)) = waiters.pop_front() {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:191]: I think this operation will execute successfully in most case.

File: src/storage/src/hummock/utils.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:196]: good catch.
I will design a guard to notify other thread when this future is canceled.

File: src/storage/src/hummock/utils.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:238]: We must try to acquire with lock, because if there is no tracker hold by other threads, we can not notify this waiter forever.

File: src/common/src/config.rs (hunk 3 above)
Comment (Little-Wallace) [reply src/common/src/config.rs]: This configuration has been removed

**Review by wenym1 — COMMENTED** [review]
Can we have some unit tests to test the correctness and performance? In unit test we can launch a multi-threaded tokio runtime and spawn many tasks to keep doing `require_memory`, `sleep` for a while and then drop the tracker, in this way we may find out some uncovered corner case.

**Review by wenym1 — COMMENTED** [review]
Rest LGTM

File: src/storage/src/hummock/utils.rs [hunk 6]
```diff
@@ -203,83 +266,82 @@ impl MemoryLimiterInner {
...
+            if first_req {
+                self.pending_request_count.store(0, AtomicOrdering::Release);
             }
-            notified.await;
+            return MemoryRequest::Ready(MemoryTrackerImpl::new(self.clone(), quota));
         }
+        let (tx, rc) = channel();
+        waiters.push_back((tx, quota));
+        // This variable can only update within lock.
+        self.pending_request_count
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:286]: Store once atomic value is lightly comparing with acquire lock for waiter queue

File: src/storage/src/hummock/utils.rs [hunk 7]
```diff
@@ -166,23 +170,82 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
-#[derive(Debug)]
+type RequestQueue = VecDeque<(Sender<MemoryTrackerImpl>, u64)>;
+enum MemoryRequest {
+    Ready(MemoryTrackerImpl),
+    Pending(Receiver<MemoryTrackerImpl>),
+}
+
+pub struct PendingRequestCancelGuard {
+    inner: Option<Arc<MemoryLimiterInner>>,
+    rc: Receiver<MemoryTrackerImpl>,
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:181]: Why?
There is no charactor 'x' in 'Receiver'.

File: src/storage/src/hummock/utils.rs [hunk 8]
```diff
@@ -166,23 +170,82 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
+}
+
+impl Drop for PendingRequestCancelGuard {
+    fn drop(&mut self) {
+        if let Some(limiter) = self.inner.take() {
+            self.rc.close();
+            if let Ok(msg) = self.rc.try_recv() {
+                drop(msg);
+                if limiter.pending_request_count.load(AtomicOrdering::Acquire) > 0 {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:190]: yes.

File: src/storage/src/hummock/utils.rs [hunk 9]
```diff
@@ -166,23 +170,82 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
+
+pub struct PendingRequestCancelGuard {
+    inner: Option<Arc<MemoryLimiterInner>>,
+    rc: Receiver<MemoryTrackerImpl>,
+}
+
+impl Drop for PendingRequestCancelGuard {
+    fn drop(&mut self) {
+        if let Some(limiter) = self.inner.take() {
+            self.rc.close();
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs]: done

File: src/storage/src/hummock/utils.rs [hunk 10]
```diff
@@ -310,41 +371,66 @@ impl MemoryLimiter {
...
-            limiter: self.inner.clone(),
-            quota,
+        match self.inner.require_memory(quota) {
+            MemoryRequest::Ready(inner) => MemoryTracker { inner },
+            MemoryRequest::Pending(rc) => {
+                let guard = PendingRequestCancelGuard {
+                    inner: Some(self.inner.clone()),
+                    rc,
+                };
+                let inner = guard.await.unwrap();
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:394]: done

File: src/storage/src/hummock/utils.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wenym1) [reply src/storage/src/hummock/utils.rs:291]: I don't get your point. Checking `first_req` does not need to acquire lock, because we still hold the lock, and we already know the value of `first_req`. Does it harm to add a check to avoid an unnecessary atomic storing?

File: src/storage/src/hummock/utils.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wenym1) [reply src/storage/src/hummock/utils.rs]: From [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch16-02-message-passing.html)
> The abbreviations tx and rx are traditionally used in many fields for transmitter and receiver respectively, so we name our variables as such to indicate each end.

Likewise, there is no 'x' in `Sender` either, but we still use `tx`.

File: src/storage/src/hummock/utils.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:291]: I means that storing once atomic value may only cost no more than 1/10 compare to calling sender

[the remark this answers is not in the document]
Comment (wenym1) [reply src/storage/src/hummock/utils.rs:291]: But it still has cost, right? We are not making a trade off between A and B, but only deciding whether A is necessary or not.

Does it hurt to add a simple check to avoid an unnecessary atomic storing? Adding a check does not increase any code complexity, and can reduce a unnecessary cost. I don't see any downside on it.

File: src/storage/src/hummock/utils.rs [hunk 11]
```diff
@@ -203,83 +271,82 @@ impl MemoryLimiterInner {
...
-                .is_ok()
-        {
-            // fast path.
-            return;
+    fn require_memory(self: &Arc<Self>, quota: u64) -> MemoryRequest {
+        let mut waiters = self.controller.lock();
+        let first_req = waiters.is_empty();
+        if first_req {
+            // When this request is the first waiter but the previous `MemoryTracker` is just release a large quota, it may skip notifying this waiter because it has checked `inflight_barrier` and found it was zero. So we must set it one and retry `try_require_memory` again to avoid deadlock.
+            self.pending_request_count.store(1, AtomicOrdering::Release);
```
Comment (hzxa21) [inline src/storage/src/hummock/utils.rs:279]: I think it is simpler and easier to understand if we make this an atomic boolean and rename it to `has_waiters`

File: src/storage/src/hummock/utils.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/src/hummock/utils.rs:291]: This line is not needed if:
- We make `pending_request_count` a atomic bool.
- We only set the atomic bool to true on first request, which we already do in this PR.
- We only set the atomic bool to false when the waiter queue becomes empty in `may_notify_waiters`.

File: src/storage/src/hummock/utils.rs (hunk 11 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:279]: ok

File: src/storage/src/hummock/utils.rs [hunk 12]
```diff
@@ -166,23 +170,88 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
+
+pub struct PendingRequestCancelGuard {
+    inner: Option<Arc<MemoryLimiterInner>>,
+    rx: Receiver<MemoryTrackerImpl>,
+}
+
+impl Drop for PendingRequestCancelGuard {
+    fn drop(&mut self) {
+        if let Some(limiter) = self.inner.take() {
+            // close rc before calling `try_recv`, it will make `MemoryTrackerImpl` which generates after dropping of `PendingRequestCancelGuard` drop in loop of `may_notify_waiters` in other thread.
```
Comment (MrCroxx) [inline src/storage/src/hummock/utils.rs:187]: A small typo.

```suggestion
            // close rx before calling `try_recv`, it will make `MemoryTrackerImpl` which generates after dropping of `PendingRequestCancelGuard` drop in loop of `may_notify_waiters` in other thread.
```

**Review by MrCroxx — APPROVED** [review]
LGTM

**Review by wenym1 — APPROVED** [review]
LGTM

File: src/storage/src/hummock/utils.rs [hunk 13]
```diff
@@ -203,83 +272,79 @@ impl MemoryLimiterInner {
...
-                )
-                .is_ok()
-        {
-            // fast path.
-            return;
+    fn require_memory(self: &Arc<Self>, quota: u64) -> MemoryRequest {
+        let mut waiters = self.controller.lock();
+        let first_req = waiters.is_empty();
+        if first_req {
+            // When this request is the first waiter but the previous `MemoryTracker` is just release a large quota, it may skip notifying this waiter because it has checked `has_waiter` and found it was false. So we must set it one and retry `try_require_memory` again to avoid deadlock.
```
Comment (Li0k) [inline src/storage/src/hummock/utils.rs:279]: typo: “set it one”

File: src/storage/src/hummock/utils.rs [hunk 14]
```diff
@@ -166,23 +170,88 @@ pub fn prune_nonoverlapping_ssts<'a>(
...
+        if !self.has_waiter.load(AtomicOrdering::Acquire) {
+            return;
+        }
+        let mut waiters = self.controller.lock();
+        while let Some((_, quota)) = waiters.front() {
+            if !self.try_require_memory(*quota) {
+                break;
+            }
+            let (tx, quota) = waiters.pop_front().unwrap();
+            let _ = tx.send(MemoryTrackerImpl::new(self.clone(), quota));
```
Comment (hzxa21) [inline src/storage/src/hummock/utils.rs:247]: The `MemoryTrackerImpl` and the newly added `PendingRequestCancelGuard` seem to be an over-design for the corner case mentioned above. Let's analyze all the cases when quota is acquired:
1. When the quota is acquired via `MemoryRequest::Ready` (L287), we need to release the quota and notify waiters.
2. When the quota is acquired here in `may_notify_waiters` in L243, there are two sub-cases:
    a. `tx.send` here succeeds, regardless of whether the rx is dropped or not, we need to release quota and notify waiter.
    b. `tx.send` here fails (due to rx has been dropped before tx.send). we only need to release quota but **not** notify watier (Otherwise, it will deadlock).


Therefore, the implementation can be greatly simplified without introducing `MemoryTrackerImpl` and `PendingRequestCancelGuard`.

```rust
// MemoryTrackerImpl is not needed.
struct MemoryTracker {
    limiter: Arc<MemoryLimiterInner>,
    quota: Option<u64>,
}

// MemoryTracker::drop will release quota + notifier waiters if quota is not None (1 + 2.a)
impl Drop for MemoryTracker {
    fn drop(&mut self) {
        if let Some(quota) = self.quota.take() {
            self.limiter.release_quota(quota);
            self.limiter.may_notify_waiters();
        }
    }
}

// Add a new method to releae quota without notify waiters.
impl MemoryTracker {
  fn release_quota(&mut self) {
      if let Some(quota) = self.quota.take() {
            self.limiter.release_quota(quota);
      }
  }
}

...

// MemoryRequest takes MemoryTracker
enum MemoryRequest {
    Ready(MemoryTracker),
    Pending(Receiver<MemoryTracker>),
}

...

impl MemoryLimiterInner {
    ...
    fn may_notify_waiters(self: &Arc<Self>) {
        ...
        let mut waiters = self.controller.lock();
        while let Some((_, quota)) = waiters.front() {
            if !self.try_require_memory(*quota) {
                break;
            }
            ...
            if let Err(SendError(tracker)) = tx.send(MemoryTracker::new(self.clone(), quota)) {
                // 2.b
                tracker.release_quota(quota);
            }
    }
    ...
}

...

impl MemoryLimiter {
    pub async fn require_memory(&self, quota: u64) -> MemoryTracker {
        // PendingRequestCancelGuard is not needed
        match self.inner.require_memory(quota) {
            MemoryRequest::Ready(tracker) => tracker,
            MemoryRequest::Pending(rx) => rx.await
    }
}
```

**Review by hzxa21 — CHANGES_REQUESTED** [review]

File: src/storage/src/hummock/utils.rs (hunk 14 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:247]: We can not send `MemoryTracker` with in lock of  `controller` because if it drops in this method, it may call `may_notify_waiters ` again, which would acquire one lock twice.

Comment (Little-Wallace) [reply src/storage/src/hummock/utils.rs:247]: We met this bug in https://github.com/risingwavelabs/risingwave/pull/6634

Comment (wenym1) [reply src/storage/src/hummock/utils.rs]: If https://github.com/tokio-rs/tokio/pull/6558 looks good from the tokio community, we can have a patch on the tokio version we are using, and then we can adopt the code in this comment.

**Comment by wenym1** [comment]
I found [loom](https://github.com/tokio-rs/loom), an interesting framework to test concurrent utilities. It provides mocked atomic variable and mutex, and `block_on` to support `future.await`, which covers all functionalities we use to sync between multiple parallelisms in our memory limiter. Shall we have some test written with the loom framework? It will be really helpful to prove the correctness of our implementation. You may refer to the [loom test](https://github.com/tokio-rs/tokio/tree/master/tokio/src/sync/tests) written for `tokio::sync`.

File: src/storage/src/hummock/utils.rs [hunk 15]
```diff
@@ -309,42 +330,41 @@ impl MemoryLimiter {
...
+        if self.limiter.try_require_memory(target - quota) {
+            self.quota = Some(target);
             true
         } else {
             false
         }
     }
 }
 
+// We must notify waiters outside `MemoryTracker` to avoid dead-lock and loop-owner.
```
Comment (Li0k) [inline src/storage/src/hummock/utils.rs:361]: typo: doc

**Review by Li0k — APPROVED** [review]
Rest LGTM
